src/Features/Core/Portable/CodeFixesAndRefactorings/CodeActionRequestPriorityProvider.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeActions
+{
+    internal abstract partial class CodeActionRequestPriorityProvider
+    {
+        /// <summary>
+        /// Default provider with <see cref="CodeActionRequestPriority.None"/> <see cref="Priority"/>.
+        /// </summary>
+        public static readonly CodeActionRequestPriorityProvider Default = DefaultProvider.Instance;
+
+        protected CodeActionRequestPriorityProvider(CodeActionRequestPriority priority)
+        {
+            Priority = priority;
+        }
+
+        public CodeActionRequestPriority Priority { get; }
+
+        protected abstract bool IsDeprioritizedAnalyzerWithLowPriority(DiagnosticAnalyzer analyzer);
+
+        /// <summary>
+        /// Tracks the given <paramref name="analyzer"/> as a de-prioritized analyzer that should be moved to
+        /// <see cref="CodeActionRequestPriority.Low"/> bucket.
+        /// </summary>
+        public abstract void AddDeprioritizedAnalyzerWithLowPriority(DiagnosticAnalyzer analyzer);
+
+        public abstract CodeActionRequestPriorityProvider With(CodeActionRequestPriority priority);
```

Comment:
still not a fan of this pseudo immutable/mutable aspect here.  will think if there's an alternate approach that would work better.   One quesion i have is why can't AddDeprioritizedAnalyzerWithLowPriority return fresh instance.

Reply:
> One quesion i have is why can't AddDeprioritizedAnalyzerWithLowPriority return fresh instance.

The core issue is the `CodeActionRequestPriorityProvider` is created all the way up the stack by AsyncSuggestedActionSource. It asks to execute `Normal` priority analyzers, but we might de-prioritize some normal ones to `Low` priority bucket. Subsequently, when it processes `Low` bucket and requests to execute `Low` priority analyzers, these de-prioritized analyzers need to be executed. This is what introduces statefulness and requires mutable `CodeActionRequestPriorityProvider` which can maintain state across `Normal` and `Low` priority calls. We can not determine this in a stateless fashion because the de-prioritization heuristic we settled upon has checks such as _If analyzer has already executed in background analysis, and has cached diagnostics, we retain the normal pri LB order for its code fixes_, and this aspect can change across `Normal` and `Low` priority calls